src/main/main.ts

```diff
@@ -347,197 +360,188 @@ ipcMain.on(ChannelEnum.SET_PREAMP, async (event, arg) => {
+  state.filters[filterId].quality = quality;
+  await handleUpdate(event, channel + filterId);
 });
 
 ipcMain.on(ChannelEnum.GET_FILTER_TYPE, async (event, arg) => {
   const channel = ChannelEnum.GET_FILTER_TYPE;
-  const filterIndex = parseInt(arg[0], 10) || 0;
+  const filterId = arg[0];
 
-  // Filter index must be within the length of the filters array
-  if (filterIndex < 0 || filterIndex >= state.filters.length) {
-    handleError(event, channel + filterIndex, ErrorCode.INVALID_PARAMETER);
+  // Filter id must exist
+  if (!checkFilterIdExistence(event, channel, filterId)) {
     return;
   }
 
   const reply: TSuccess<string> = {
-    result: state.filters[filterIndex].type,
+    result: state.filters[filterId].type,
   };
-  event.reply(channel + filterIndex, reply);
+  event.reply(channel + filterId, reply);
 });
 
 ipcMain.on(ChannelEnum.SET_FILTER_TYPE, async (event, arg) => {
   const channel = ChannelEnum.SET_FILTER_TYPE;
-  const filterIndex = parseInt(arg[0], 10) || 0;
+  const filterId = arg[0];
   const filterType = arg[1];
 
-  // Filter index must be within the length of the filters array
-  if (filterIndex < 0 || filterIndex >= state.filters.length) {
-    handleError(event, channel + filterIndex, ErrorCode.INVALID_PARAMETER);
+  // Filter id must exist
+  if (!checkFilterIdExistence(event, channel, filterId)) {
     return;
   }
 
   if (!Object.values(FilterTypeEnum).includes(filterType)) {
-    handleError(event, channel + filterIndex, ErrorCode.INVALID_PARAMETER);
+    handleError(event, channel + filterId, ErrorCode.INVALID_PARAMETER);
     return;
   }
 
-  state.filters[filterIndex].type = filterType as FilterTypeEnum;
-  await handleUpdate(event, channel + filterIndex);
+  state.filters[filterId].type = filterType as FilterTypeEnum;
+  await handleUpdate(event, channel + filterId);
 });
 
 ipcMain.on(ChannelEnum.GET_FILTER_COUNT, async (event) => {
   const reply: TSuccess<number> = {
-    result: state.filters.length,
+    result: Object.keys(state.filters).length,
   };
   event.reply(ChannelEnum.GET_FILTER_COUNT, reply);
 });
 
 ipcMain.on(ChannelEnum.ADD_FILTER, async (event, arg) => {
   const channel = ChannelEnum.ADD_FILTER;
-  const insertIndex: number = arg[0];
+  const frequency: number = arg[0];
 
   // Cannot exceed the maximum number of filters
-  if (state.filters.length === MAX_NUM_FILTERS) {
+  // Frequency must be in valid range
+  if (
+    Object.keys(state.filters).length === MAX_NUM_FILTERS ||
+    frequency < MIN_FREQUENCY ||
+    frequency > MAX_FREQUENCY
+  ) {
     handleError(event, channel, ErrorCode.INVALID_PARAMETER);
     return;
   }
 
-  const frequency = computeAvgFreq(state.filters, insertIndex);
   const filterId = uid(8);
-  state.filters
-    .splice(insertIndex, 0, {
-      id: filterId,
-      frequency,
-      gain: 0,
-      quality: 1,
-      type: FilterTypeEnum.PK,
-    })
-    .sort(sortHelper);
+  state.filters[filterId] = {
+    ...getDefaultFilterWithId(),
```

Comment:
instead of calling `uid()` twice (once in line 519, once in `getDefaultFilterWithId()`), why don't we just store the filter returned from `getDefaultFilterWithId()` and then use the id from there to add the filter to the state?